### PR TITLE
ad796x: Add project and driver for ad796x-fmc

### PR DESCRIPTION
**Pull request by ahaslam2**
This adds driver files and a project test the ad796x series of pulsar Differential ADC.

The HDL for this project is still not merged, and probably its why CI is failing.
perhaps review can start without the HDL 

this was tested using the HDL binary at https://github.com/adi-ses/sea-misc/tree/main/AD796x-PRJ

## PR Type
- [ ] Bug fix (change that fixes an issue)
- [x] New feature (change that adds new functionality)
- [ ] Breaking change (has dependencies in other repos or will cause CI to fail)

## PR Checklist
- [x] I have followed the [Coding style guidelines](http://analogdevicesinc.github.io/no-OS/drivers_guide.html#coding-style)
- [x] I have performed a self-review of the changes
- [x] I have commented my code, at least hard-to-understand parts
- [x] I have build all projects affected by the changes in this PR
- [x] I have tested in hardware affected projects, at the relevant boards
- [x] I have signed off all commits from this PR
- [ ] I have updated the documentation (wiki pages, ReadMe etc), if applies

### CODE REVIEW

File: drivers/adc/ad796x/ad796x.c [hunk 1]
```diff
@@ -0,0 +1,168 @@
...
+/******************************************************************************/
+#include <stdlib.h>
+#include <errno.h>
+#include "ad796x.h"
+#include "no_os_alloc.h"
+
+/******************************************************************************/
+/************************ Functions Definitions *******************************/
+/******************************************************************************/
+/******************************************************************************/
```
Comment (rbolboac) [inline drivers/adc/ad796x/ad796x.c:51]: remove extra comment line

File: drivers/adc/ad796x/ad796x.c [hunk 2]
```diff
@@ -0,0 +1,168 @@
...
+
```
Comment (rbolboac) [inline drivers/adc/ad796x/ad796x.c:52]: remove extra blank line

File: drivers/adc/ad796x/ad796x.h [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+#define __AD796X_H__
+
+/******************************************************************************/
+/***************************** Include Files **********************************/
+/******************************************************************************/
+#include "no_os_gpio.h"
+
+/******************************************************************************/
+/********************** Macros and Constants Definitions **********************/
+/******************************************************************************/
```
Comment (rbolboac) [inline drivers/adc/ad796x/ad796x.h:49]: remove this seeing how there are no macros or constants definitions

File: drivers/adc/ad796x/ad796x.h [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+ * @struct ad796x_dev
+ * @brief AD796X Device structure.
+ */
+struct ad796x_dev {
+	/** Device communication descriptor */
+	struct no_os_gpio_desc		*gpio_adc_en3_fmc;
+	struct no_os_gpio_desc		*gpio_adc_en2_fmc;
+	struct no_os_gpio_desc		*gpio_adc_en1_fmc;
+	struct no_os_gpio_desc		*gpio_adc_en0_fmc;
+
```
Comment (rbolboac) [inline drivers/adc/ad796x/ad796x.h:77]: remove extra blank line

**Review by rbolboac — COMMENTED** [review]
some remarks from my side, also:
please create a different commit for the driver and in the same commit please also add the driver documentation
please use the new project structure

**Comment by ahaslam2** [comment]
@rbolboac 

ok ill separate driver and project. but 
what do you mean by "new project structure"?

**Comment by rbolboac** [comment]
> @rbolboac
> 
> ok ill separate driver and project. but what do you mean by "new project structure"?

By new project structure I mean the src folder should have platform independent examples, with common folder and platform specific folders. This is an example from iio_demo project:
```
├── src
│   ├── common
│   │   ├── app_config.h
│   │   ├── common_data.c
│   │   └── common_data.h
│   ├── examples
│   │   ├── examples_src.mk
│   │   ├── iio_example
│   │   │   ├── iio_example.c
│   │   │   └── iio_example.h
│   │   ├── iio_sw_trigger_example
│   │   │   ├── iio_sw_trigger_example.c
│   │   │   └── iio_sw_trigger_example.h
│   │   └── iio_timer_trigger_example
│   │       ├── iio_timer_trigger_example.c
│   │       └── iio_timer_trigger_example.h
│   └── platform
│       ├── aducm3029
│       │   ├── main.c
│       │   ├── parameters.c
│       │   ├── parameters.h
│       │   └── platform_src.mk
│       ├── linux
│       │   ├── main.c
│       │   ├── parameters.c
│       │   ├── parameters.h
│       │   └── platform_src.mk
│       ├── maxim
│       │   ├── main.c
│       │   ├── parameters.c
│       │   ├── parameters.h
│       │   └── platform_src.mk
│       ├── pico
│       │   ├── main.c
│       │   ├── parameters.c
│       │   ├── parameters.h
│       │   └── platform_src.mk
│       ├── platform_includes.h
│       ├── stm32
│       │   ├── main.c
│       │   ├── parameters.c
│       │   ├── parameters.h
│       │   └── platform_src.mk
│       └── xilinx
│           ├── main.c
│           ├── parameters.c
│           ├── parameters.h
│           └── platform_src.mk

```

**Comment by ahaslam2** [comment]
Ok, i did not know this structure was a requirement, will do.

**Comment by ahaslam2** [comment]
one question @rbolboac 

i would suppose the structure requirement is to keep the examples platform
independant? 

i see things like axi_dac_init_begin(&phy->tx_dac, &tx_dac_init);
that i think are tied to some specific fpga/platform ?

https://github.com/analogdevicesinc/no-OS/blob/main/projects/adrv902x/src/examples/iio_example/iio_example.c#L304

are there some rules as to what goes into platform/common/example?

**Comment by rbolboac** [comment]
well, the idea is to have the examples generic and platform independent, but if that is not possible I suppose it is alright.
in that case adrv902x project can be used as an example.

**Comment by ahaslam2** [comment]
The failure of CI is currently due to the missing HDL:

```test -d latest_build/ad796x_fmcz
 -> Building ad796x_fmcz (iio) -- xilinx -- ad796x_fmc_zed
 -> cp /no-OS_builds/builds_main/new_hardware/ad796x_fmc_zed/system_top.xsa /no-OS_builds/builds_main/hardware/ad796x_fmc_zed.xsa
 -> ERROR
 -> See log logs/ad796x_fmcz_xilinx_iio_ad796x_fmc_zed.txt -- Use cat (linux) or type (windows) to see colored output
Error occured: 1
````

File: drivers/adc/ad796x/ad796x.c [hunk 5]
```diff
@@ -0,0 +1,297 @@
...
+ * @param buf - The buffer to fill.
+ * @param samples - number of samples to read.
+ * @return 0 in case of success, negative error code otherwise.
+ */
+int32_t ad796x_read_data(struct ad796x_dev *dev,
+			 uint32_t *buf,
+			 uint16_t samples)
+{
+	int ret;
+
```
Comment (rbolboac) [inline drivers/adc/ad796x/ad796x.c:85]: blank line can be removed, no reason to leave blank lines between declarations. see everywhere where it applies

File: drivers/adc/ad796x/ad796x.c [hunk 6]
```diff
@@ -0,0 +1,297 @@
...
+	int ret, i;
+
+	struct no_os_gpio_desc *desc[] = {
+		dev->gpio_adc_en3_fmc,
+		dev->gpio_adc_en2_fmc,
+		dev->gpio_adc_en1_fmc,
+		dev->gpio_adc_en0_fmc,
+	};
+
+	for (i = 0; i < 4; i++) {
```
Comment (rbolboac) [inline drivers/adc/ad796x/ad796x.c:119]: create a macro value for that magic value

File: drivers/adc/ad796x/ad796x.c [hunk 7]
```diff
@@ -0,0 +1,297 @@
...
+	ret = axi_clkgen_init(&dev->clkgen, init_param->clkgen_ip);
+	if (ret) {
+		pr_err("error: axi_clkgen_init error: %d\n", ret);
+		goto err_clkgen;
+	}
+
+	ret = axi_clkgen_set_rate(dev->clkgen, 125000000);
+	if (ret) {
+		pr_err("error: axi_clkgen_set_rate error: %d\n", ret);
+		goto err_set_rate;
```
Comment (rbolboac) [inline drivers/adc/ad796x/ad796x.c:213]: no need for an extra label, can use err_clkgen

File: drivers/adc/ad796x/ad796x.c [hunk 8]
```diff
@@ -0,0 +1,297 @@
...
+	no_os_pwm_remove(dev->axi_pwm_1);
+err_pwm_1:
+	no_os_pwm_remove(dev->axi_pwm_0);
+err_pwm_0:
+err_set_rate:
+	axi_clkgen_remove(dev->clkgen);
+err_clkgen:
+	ad796x_gpio_remove(dev);
+error_gpio:
+	no_os_free(dev);
```
Comment (rbolboac) [inline drivers/adc/ad796x/ad796x.c:256]: I think it is safe to have a single error label (but please check before).
basically, if the init function has not been called, the desc will be null and the remove functions should check whether the desc is null or not. if it is null nothing will happen, if it is not null it will be freed.

File: drivers/adc/ad796x/ad796x.h [hunk 9]
```diff
@@ -0,0 +1,119 @@
...
+/******************************************************************************/
+
+/* Initialize the device. */
+int ad796x_init(struct ad796x_dev **device,
+		struct ad796x_init_param *init_param);
+
+/* Remove the device and release resources. */
+int ad796x_remove(struct ad796x_dev *dev);
+
+int32_t ad796x_read_data(struct ad796x_dev *dev, uint32_t *buf,
```
Comment (rbolboac) [inline drivers/adc/ad796x/ad796x.h:115]: missing doxygen description

File: projects/ad796x_fmcz/builds.json [hunk 10]
```diff
@@ -0,0 +1,10 @@
+{
+  "xilinx": {
+    "iio": {
+      "flags": "IIO_EXAMPLE=y BASIC_EXAMPLE=n",
+      "hardware": [
+        "ad796x_fmc_zed"
```
Comment (rbolboac) [inline projects/ad796x_fmcz/builds.json:6]: the build is probably failing because of this... the hdl project is here from whay I have been tols: https://github.com/analogdevicesinc/hdl/tree/main/projects/pulsar_lvds_adc
so the hardware artifact is actually named pulsar_lvds_adc_zed
let's give this a try and see if the build is successful

File: projects/ad796x_fmcz/readme.txt [hunk 11]
```diff
@@ -0,0 +1 @@
+TODO
```
Comment (rbolboac) [inline projects/ad796x_fmcz/readme.txt:1]: remove this file from this commit if it is not useful

File: projects/ad796x_fmcz/src/examples/basic_example/basic_example.c [hunk 12]
```diff
@@ -0,0 +1,95 @@
...
+*******************************************************************************/
+
+/******************************************************************************/
+/***************************** Include Files **********************************/
+/******************************************************************************/
+#include "basic_example.h"
+#include "common_data.h"
+#include "ad796x.h"
+#include "no_os_util.h"
+
```
Comment (rbolboac) [inline projects/ad796x_fmcz/src/examples/basic_example/basic_example.c:47]: remove extra blank line

File: projects/ad796x_fmcz/src/examples/basic_example/basic_example.c [hunk 13]
```diff
@@ -0,0 +1,95 @@
...
+
+/******************************************************************************/
+/************************ Functions Definitions *******************************/
+/******************************************************************************/
+/***************************************************************************//**
```
Comment (rbolboac) [inline projects/ad796x_fmcz/src/examples/basic_example/basic_example.c:52]: add blank line in between

File: projects/ad796x_fmcz/src/examples/basic_example/basic_example.c [hunk 14]
```diff
@@ -0,0 +1,95 @@
...
+ * @brief BASIC example main execution.
+ *
+ * @return ret - Result of the example execution. If working correctly, will
+ *               execute continuously function basic_app_run and will not return.
+*******************************************************************************/
+int basic_example_main()
+{
+
```
Comment (rbolboac) [inline projects/ad796x_fmcz/src/examples/basic_example/basic_example.c:60]: remove blank line

File: projects/ad796x_fmcz/src/examples/basic_example/basic_example.c [hunk 15]
```diff
@@ -0,0 +1,95 @@
...
+	uint32_t *buf = (uintptr_t)ADC_DDR_BASEADDR;
+	int i, ret;
+
```
Comment (rbolboac) [inline projects/ad796x_fmcz/src/examples/basic_example/basic_example.c:63]: remove blank line

File: projects/ad796x_fmcz/src/examples/basic_example/basic_example.c [hunk 16]
```diff
@@ -0,0 +1,95 @@
...
+	printf("\n init\n");
+
+	ret = ad796x_init(&adc_dev, &ad796x_init_param);
+	if (ret)
+		return ret;
+	printf("\n Capture start.\n");
+
+	ret = ad796x_read_data(adc_dev, buf, SAMPLES_PER_CHANNEL);
+	if (ret)
+		return ret;
```
Comment (rbolboac) [inline projects/ad796x_fmcz/src/examples/basic_example/basic_example.c:75]: ad796x_remove should be called before returning to avoid memory leaks and also an error message might be useful

File: projects/ad796x_fmcz/src/examples/basic_example/basic_example.c [hunk 17]
```diff
@@ -0,0 +1,95 @@
...
+
+	for (i = 0; i < SAMPLES_PER_CHANNEL; i++, buf++)
+		printf("CH1: %ld\n", *buf);
+
+	printf("\n Capture done.\n");
+
+	return 0;
```
Comment (rbolboac) [inline projects/ad796x_fmcz/src/examples/basic_example/basic_example.c:82]: ad796x_remove should be called before returning to avoid memory leaks

File: projects/ad796x_fmcz/src/examples/basic_example/basic_example.c [hunk 18]
```diff
@@ -0,0 +1,95 @@
...
+
+
+
+
+
+
+
+
+
+
```
Comment (rbolboac) [inline projects/ad796x_fmcz/src/examples/basic_example/basic_example.c:95]: remove all extra blank lines

File: projects/ad796x_fmcz/src/examples/basic_example/basic_example.h [hunk 19]
```diff
@@ -0,0 +1,51 @@
...
+ * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+*******************************************************************************/
+#ifndef __BASIC_EXAMPLE_H__
+#define __BASIC_EXAMPLE_H__
+
+/******************************************************************************/
+/***************************** Include Files **********************************/
+/******************************************************************************/
+
```
Comment (rbolboac) [inline projects/ad796x_fmcz/src/examples/basic_example/basic_example.h:45]: remove this seeing how there are no includes

File: projects/ad796x_fmcz/src/examples/examples_src.mk [hunk 20]
```diff
@@ -0,0 +1,21 @@
+ifeq (y,$(strip $(IIO_EXAMPLE)))
+CFLAGS += -DIIO_EXAMPLE=1
+SRCS += $(PROJECT)/src/examples/iio_example/iio_example.c
+INCS += $(PROJECT)/src/examples/iio_example/iio_example.h
+IIOD=y
+SRC_DIRS += $(NO-OS)/iio/iio_app
+endif
+
```
Comment (rbolboac) [inline projects/ad796x_fmcz/src/examples/examples_src.mk:8]: remove extra blank line

File: projects/ad796x_fmcz/src/examples/iio_example/iio_example.c [hunk 21]
```diff
@@ -0,0 +1,123 @@
...
+#include "common_data.h"
+#include "ad796x.h"
+#include "no_os_util.h"
+#include "iio_app.h"
+#include "iio_axi_adc.h"
+
+/******************************************************************************/
+/************************ Functions Definitions *******************************/
+/******************************************************************************/
+/***************************************************************************//**
```
Comment (rbolboac) [inline projects/ad796x_fmcz/src/examples/iio_example/iio_example.c:53]: add a blank line in between

File: projects/ad796x_fmcz/src/examples/iio_example/iio_example.c [hunk 22]
```diff
@@ -0,0 +1,123 @@
...
+		.rx_adc = adc_dev->ad796x_core,
+		.rx_dmac = adc_dev->axi_dmac,
+		.scan_type_common = &init_scan_type,
+		.dcache_invalidate_range = (void (*)(uint32_t,
+						     uint32_t))Xil_DCacheInvalidateRange,
+	};
+
+	ret = iio_axi_adc_init(&iio_axi_adc_desc, &iio_axi_adc_init_par);
+	if (ret)
+		return ret;
```
Comment (rbolboac) [inline projects/ad796x_fmcz/src/examples/iio_example/iio_example.c:103]: ad796x_remove should be called in case of an error

File: projects/ad796x_fmcz/src/examples/iio_example/iio_example.c [hunk 23]
```diff
@@ -0,0 +1,123 @@
...
+			       dev_desc, &adc_buff, NULL, NULL),
+	};
+
+	app_init_param.devices = devices;
+	app_init_param.nb_devices = NO_OS_ARRAY_SIZE(devices);
+	app_init_param.uart_init_params = iio_uart_ip;
+
+	ret = iio_app_init(&app, app_init_param);
+	if (ret)
+		return ret;
```
Comment (rbolboac) [inline projects/ad796x_fmcz/src/examples/iio_example/iio_example.c:118]: the according remove functions should be called in case of an error

File: projects/ad796x_fmcz/src/examples/iio_example/iio_example.c [hunk 24]
```diff
@@ -0,0 +1,123 @@
...
+
+	return iio_app_run(app);
```
Comment (rbolboac) [inline projects/ad796x_fmcz/src/examples/iio_example/iio_example.c:120]: the according remove functions should be called in case of an error

File: projects/ad796x_fmcz/src/examples/iio_example/iio_example.h [hunk 25]
```diff
@@ -0,0 +1,51 @@
...
+ * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+*******************************************************************************/
+#ifndef __IIO_EXAMPLE_H__
+#define __IIO_EXAMPLE_H__
+
+/******************************************************************************/
+/***************************** Include Files **********************************/
+/******************************************************************************/
+
```
Comment (rbolboac) [inline projects/ad796x_fmcz/src/examples/iio_example/iio_example.h:45]: remove this

File: projects/ad796x_fmcz/src/platform/xilinx/parameters.c [hunk 26]
```diff
@@ -0,0 +1,55 @@
...
+#ifdef XPAR_XUARTLITE_NUM_INSTANCES
+	.type = UART_PL,
+#else
+	.type = UART_PS,
+	.irq_id = UART_IRQ_ID
+#endif
+};
+
+
+
```
Comment (rbolboac) [inline projects/ad796x_fmcz/src/platform/xilinx/parameters.c:55]: remove extra blank lines

File: projects/ad796x_fmcz/src/platform/xilinx/parameters.h [hunk 27]
```diff
@@ -0,0 +1,112 @@
...
+#define __PARAMETERS_H__
+
+/******************************************************************************/
+/***************************** Include Files **********************************/
+/******************************************************************************/
+#include <xparameters.h>
+#include <xil_cache.h>
+#include <xilinx_uart.h>
+#include "xilinx_gpio.h"
+
```
Comment (rbolboac) [inline projects/ad796x_fmcz/src/platform/xilinx/parameters.h:50]: remove extra blank line

File: projects/ad796x_fmcz/src/platform/xilinx/parameters.h [hunk 28]
```diff
@@ -0,0 +1,112 @@
...
+#ifdef XPS_BOARD_ZCU102
+#define UART_IRQ_ID		XPAR_XUARTPS_0_INTR
+#else
+#define UART_IRQ_ID		XPAR_XUARTPS_1_INTR
+#endif
+
+#else // _XPARAMETERS_PS_H_
+#define UART_DEVICE_ID	XPAR_AXI_UART_DEVICE_ID
+#define INTC_DEVICE_ID	XPAR_INTC_SINGLE_DEVICE_ID
+#define UART_IRQ_ID		XPAR_AXI_INTC_AXI_UART_INTERRUPT_INTR
```
Comment (rbolboac) [inline projects/ad796x_fmcz/src/platform/xilinx/parameters.h:68]: use the same indentation

File: projects/ad796x_fmcz/src/platform/xilinx/parameters.h [hunk 29]
```diff
@@ -0,0 +1,112 @@
...
+
+/* This value can be modified based on the number
+of samples needed to be stored in the device buffer
+and based on the available RAM memory of the platform */
+#define SAMPLES_PER_CHANNEL_PLATFORM    20000
+#define UART_BAUDRATE	                115200
+
+#define MAX_SIZE_BASE_ADDR		(SAMPLES_PER_CHANNEL_PLATFORM * 4)
+
+
```
Comment (rbolboac) [inline projects/ad796x_fmcz/src/platform/xilinx/parameters.h:98]: remove extra blank line

**Review by rbolboac — COMMENTED** [review]
some remarks from my side

File: drivers/adc/ad796x/ad796x.c (hunk 8 above)
Comment (ahaslam2) [reply drivers/adc/ad796x/ad796x.c:256]: Ok, that just means that the convention is that  if the desc returns something that
something has to be freed (even if the int func fails). 

i guess i can add those checks in remove and just call remove then.

Comment (rbolboac) [reply drivers/adc/ad796x/ad796x.c:256]: Well... the convention is that if something is allocated and ret is not 0, it should be freed before returning.
However, seeing how we use calloc, all pointers are initialized with null, so if we call the remove function, even if memory has not been allocated, then the remove function should check for null and free the memory only if the pointer is not null.
Assuming that all remove functions check for null pointer before attempting to free it, it is safe to call the remove function even if the init has not been called. This will make the code easier to follow and shorter (a single label can be used).

File: projects/ad796x_fmcz/builds.json (hunk 10 above)
Comment (ahaslam2) [reply projects/ad796x_fmcz/builds.json:6]: i tested using prebuilt a HDL shared sea-misc. i dont think this was the pulsar_lvds
as the readme in there does not mention anything about pulsar. Also the defines in parameters
do not not mention PULSAR. 

I think the HDL for this part still needs to be merged. i will ask

File: drivers/adc/ad796x/ad796x.c (hunk 8 above)
Comment (ahaslam2) [reply drivers/adc/ad796x/ad796x.c:256]: RE: if something is allocated and ret is not 0, it should be freed before returning

What i see from the no os-code is a bit confusing:

axi_dmac does not free before returning -1
(https://github.com/analogdevicesinc/no-OS/blob/main/drivers/axi_core/axi_dmac/axi_dmac.c#L349)

axi_adc_core frees the allocation before returning -1 But the the pointer is not set to null after the free. 
https://github.com/analogdevicesinc/no-OS/blob/main/drivers/axi_core/axi_adc_core/axi_adc_core.c#L677


So in case one of the int function call fails, im not sure we can trust the pointer to be freed, or
to be freed and null.

Comment (rbolboac) [reply drivers/adc/ad796x/ad796x.c:256]: regarding https://github.com/analogdevicesinc/no-OS/blob/main/drivers/axi_core/axi_dmac/axi_dmac.c#L349, it looks like the remove function should be called if ret is not 0. so the suggestion works great in this case. I do not know why this implementation was done like that. personally I would rework it into something like:
```
struct axi_dmac *dmac;

dmac = (struct axi_dmac *)no_os_calloc(1, sizeof(*dmac));
if (!dmac)
	return -1;

dmac->name = init->name;
dmac->base = init->base;
dmac->irq_option = init->irq_option;

int32_t status = axi_dmac_detect_caps(dmac);
if (status < 0)
	goto free;

*dmac_core = dmac;

return 0;

free:
	no_os_free(dmac);
	return -1;
```

regarding https://github.com/analogdevicesinc/no-OS/blob/main/drivers/axi_core/axi_adc_core/axi_adc_core.c#L677,
the descriptor remains null in case of an error, only in case of success the descriptor is set: https://github.com/analogdevicesinc/no-OS/blob/main/drivers/axi_core/axi_adc_core/axi_adc_core.c#L673

so basically, dev->ad796x_core will be null if init is not successful, and when calling the remove function nothing will happen. 
the problem I see is that axi_adc_remove does not check for null pointer, we should add that check I think.

Comment (ahaslam2) [reply drivers/adc/ad796x/ad796x.c:256]: yes, ok. with those changes. 

BTW i have done it this way because the logic i was following was
to keep balanced number of init/remove. 

this would mean that:
- init will free its resources in case of fail, a failed init is not a init  (like you mention)
- remove should not be called if init was not called before
- remove should not be called for a failed init
- should not assume the state of the pointer

**Comment by ahaslam2** [comment]
changes:

- use remove function instead of err_ lables in driver init error path
- call driver remove in case of error on project main  
- misc style fixes.

File: drivers/adc/ad796x/ad796x.c (hunk 8 above)
Comment (rbolboac) [reply drivers/adc/ad796x/ad796x.c]: I understand... this is not mandatory. in my case the code would look cleaner...
also, to me there is no real value in adding a duplicate label: `err_pwm_0`

Comment (rbolboac) [reply drivers/adc/ad796x/ad796x.c]: also, regarding my previous comment `the problem I see is that axi_adc_remove does not check for null pointer, we should add that check I think.` there is no problem with that API, it is safe to call free for null pointers.

File: drivers/adc/ad796x/ad796x.c [hunk 30]
```diff
@@ -0,0 +1,297 @@
...
+/**
+ * @brief Remove the device and release resources.
+ * @param dev - The device structure.
+ * @return 0 in case of success, negative error code otherwise.
+ */
+int ad796x_remove(struct ad796x_dev *dev)
+{
+	int ret;
+
+	if (dev->ad796x_core) {
```
Comment (rbolboac) [inline drivers/adc/ad796x/ad796x.c:265]: the only thing I would check for null here is dev.
freeing a null pointer is allowed

File: drivers/adc/ad796x/ad796x.c [hunk 31]
```diff
@@ -0,0 +1,297 @@
...
+ * @brief Remove the device gpio resources.
+ * @param dev - The device structure.
+ * @return 0 in case of success, negative error code otherwise.
+ */
+static int ad796x_gpio_remove(struct ad796x_dev *dev)
+{
+	int ret, i;
+
+	struct no_os_gpio_desc *desc[] = {
+		dev->gpio_adc_en3_fmc,
```
Comment (rbolboac) [inline drivers/adc/ad796x/ad796x.c:120]: dev should be checked for null before trying to access its members

File: drivers/adc/ad796x/ad796x.c [hunk 32]
```diff
@@ -0,0 +1,297 @@
...
+		dev->gpio_adc_en2_fmc,
+		dev->gpio_adc_en1_fmc,
+		dev->gpio_adc_en0_fmc,
+	};
+
+	for (i = 0; i < AD796X_NUM_GPIOS; i++) {
+		if (!desc[i])
+			continue;
```
Comment (rbolboac) [inline drivers/adc/ad796x/ad796x.c:128]: no_os_gpio_remove already checks for null pointer, no need to do it here

File: drivers/adc/ad796x/ad796x.c (hunk 31 above)
Comment (ahaslam2) [reply drivers/adc/ad796x/ad796x.c:120]: ill use a table index instead to loop though the gpios.

File: drivers/adc/ad796x/ad796x.c [hunk 33]
```diff
@@ -0,0 +1,260 @@
...
+
+	ret = ad796x_gpio_remove(dev);
+	if (ret)
+		return ret;
+
+	no_os_free(dev);
+
+	return ret;
+}
+
```
Comment (amiclaus) [inline drivers/adc/ad796x/ad796x.c:260]: drop extra line

File: drivers/adc/ad796x/ad796x.h [hunk 34]
```diff
@@ -0,0 +1,116 @@
...
+int ad796x_init(struct ad796x_dev **device,
+		struct ad796x_init_param *init_param);
+
+/* Remove the device and release resources. */
+int ad796x_remove(struct ad796x_dev *dev);
+
+/* Read samples from the device */
+int32_t ad796x_read_data(struct ad796x_dev *dev, uint32_t *buf,
+			 uint16_t samples);
+
```
Comment (amiclaus) [inline drivers/adc/ad796x/ad796x.h:114]: drop extra line

**Review by amiclaus — COMMENTED** [review]
minor comments, otherwise lgtm

File: projects/ad796x_fmcz/src/examples/iio_example/iio_example.c [hunk 35]
```diff
@@ -0,0 +1,125 @@
...
+
+	ret = iio_app_init(&app, app_init_param);
+	if (ret)
+		goto err_app_init;
+
+	ret = iio_app_run(app);
+	if (ret)
+		goto err_app_run;
+
+	return 0;
```
Comment (rbolboac) [inline projects/ad796x_fmcz/src/examples/iio_example/iio_example.c:116]: I would remove these lines...if fore some reason iio_app_run returns (either with 0 or another value), the remove APIs should call before exiting this example.

File: projects/ad796x_fmcz/src/examples/iio_example/iio_example.c [hunk 36]
```diff
@@ -0,0 +1,125 @@
...
+
+err_app_run:
+	iio_app_remove(app);
+err_app_init:
+	iio_axi_adc_remove(iio_axi_adc_desc);
+err_adc_init:
+	ad796x_remove(adc_dev);
+	return ret;
```
Comment (rbolboac) [inline projects/ad796x_fmcz/src/examples/iio_example/iio_example.c:124]: if (ret) I would print an error message

File: projects/ad796x_fmcz/src/platform/xilinx/main.c [hunk 37]
```diff
@@ -0,0 +1,72 @@
...
+
+	/* Enable the instruction cache. */
+	Xil_ICacheEnable();
+	/* Enable the data cache. */
+	Xil_DCacheEnable();
+#ifdef IIO_EXAMPLE
+	ret = iio_example_main();
+#elif BASIC_EXAMPLE
+	ret = basic_example_main();
+#endif
```
Comment (rbolboac) [inline projects/ad796x_fmcz/src/platform/xilinx/main.c:70]: maybe add an `#error` message in case no example is defined

**Review by rbolboac — COMMENTED** [review]
some remarks from my side, also ad796x driver documentation is missing
same remark regarding commit message length, can have up to 72 characters per line.

File: projects/ad796x_fmcz/src/examples/iio_example/iio_example.c (hunk 36 above)
Comment (ahaslam2) [reply projects/ad796x_fmcz/src/examples/iio_example/iio_example.c:124]: Ok ill put print in after the ret that fails were i can show who produced the error.

**Comment by rbolboac** [comment]
do you plan on adding the driver documentation in this PR or in a new one?

**Comment by ahaslam2** [comment]
> do you plan on adding the driver documentation in this PR or in a new one?

will do.

File: drivers/adc/ad796x/README.rst [hunk 38]
```diff
@@ -0,0 +1,218 @@
...
+This board operates in conjunction with the Zedboard. The evaluation software
+is provided to enable the user to perform detailed analysis of the AD7960's
+performance. The technical user guide includes the detailed description of
+the operation and set up of the evaluation board and software when operated
+with the System Demonstration Platform board.
+
+
+AD796X Device Configuration
+----------------------------
+
```
Comment (rbolboac) [inline drivers/adc/ad796x/README.rst:24]: looks like device configuration documentation is missing...
in general, I think it will be useful to also describe the available APIs, just in case someone checks the documentation before the code... there are various examples here: http://analogdevicesinc.github.io/no-OS/drivers_doc.html

Comment (ahaslam2) [reply drivers/adc/ad796x/README.rst:24]: i took the example from https://analogdevicesinc.github.io/no-OS/drivers/adm1177.html
where "driver initialization " is a subtopic of "device configuration "

in fact there is no device configuration for this driver. only driver initialization and reading data,
which i mention, but perhaps not clearly enough?

Comment (rbolboac) [reply drivers/adc/ad796x/README.rst:24]: It is not clear that driver initialization is a subtopic of device configuration. they are using the same indentation. use `^^^^`, instead of `----` for driver initialization if that is the case

Comment (ahaslam2) [reply drivers/adc/ad796x/README.rst:24]: ok , i  just followed what was done in the other docs: 

```
AD405X Device Configuration
---------------------------

Driver Initialization
---------------------

In order to be able to use the device, you will have to provide the support for
the communication protocol (SPI) as well as 3 external GPIOs for the CNV pin and two
```

```
ADM1177 Device Configuration
----------------------------

Driver Initialization
---------------------

In order to be able to use the device, you will have to provide the support for
the communication protocol (I2C) as mentioned above.

The first API to be called is **adm1177_init**. Make sure that it returns 0,
which means that the driver was initialized correctly.

```

**Review by rbolboac — APPROVED** [review]

**Review by amiclaus — APPROVED** [review]
